File: 2020/15/solution.py
import os
from typing import Dict, List

def playGame(numbers: List[int], turns: int) -> int:
  history: Dict[int, int] = {}
  for i, n in enumerate(numbers):
    history[n] = i + 1
  nextNumber = 0
  for i in range(len(numbers), turns - 1):
    if nextNumber in history:
      temp = i - history[nextNumber] + 1
    else:
      temp = 0
    history[nextNumber] = i + 1
    nextNumber = temp
  return nextNumber

if __name__ == "__main__":
  with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'input.txt'),'r') as file:
    line = file.readline()
  numbers = [int(n) for n in line.split(',')]
  print(playGame(numbers, 2020))
  print(playGame(numbers, 30000000))

File: 2020/15/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split(",")
  .map(Number);

// const gameHistory = new Map();

// input.forEach((num, i) => gameHistory.set(num, i + 1));

// let turn = input.length + 1;
// let currentNumber = 0;

// while (turn < 2020) {
//   let nextNumber = 0;
//   if (gameHistory.has(currentNumber)) {
//     nextNumber = turn - gameHistory.get(currentNumber);
//   }
//   gameHistory.set(currentNumber, turn);
//   currentNumber = nextNumber;
//   turn++;
// }

// console.log(currentNumber);

// while (turn < 30000000) {
//   let nextNumber = 0;
//   if (gameHistory.has(currentNumber)) {
//     nextNumber = turn - gameHistory.get(currentNumber);
//   }
//   gameHistory.set(currentNumber, turn);
//   currentNumber = nextNumber;
//   turn++;
// }

// console.log(currentNumber);

function playMemoryGame(input, turns) {
  const gameHistory = new Map();

  input.forEach((num, i) => gameHistory.set(num, i + 1));

  let turn = input.length + 1;
  let currentNumber = 0;

  while (turn < turns) {
    let nextNumber = 0;
    if (gameHistory.has(currentNumber)) {
      nextNumber = turn - gameHistory.get(currentNumber);
    }
    gameHistory.set(currentNumber, turn);
    currentNumber = nextNumber;
    turn++;
  }
  return currentNumber;
}

console.log(playMemoryGame(input, 2020));
console.log(playMemoryGame(input, 30000000));
